File: leetcode/1081. Smallest Subsequence of Distinct Characters/main.go
package main

func smallestSubsequence(s string) string {
	lastOccurence := make([]int, 26)
	for i := range len(s) {
		lastOccurence[s[i]-'a'] = i
	}

	seen := make([]bool, 26)
	stack := []byte{}

	for i := range len(s) {
		c := s[i]
		idx := c - 'a'

		if seen[idx] {
			continue
		}

		for len(stack) > 0 && c < stack[len(stack)-1] && lastOccurence[stack[len(stack)-1]-'a'] > i {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			seen[top-'a'] = false
		}

		stack = append(stack, c)
		seen[idx] = true
	}

	return string(stack)
}

File: leetcode/1081. Smallest Subsequence of Distinct Characters/main.ts
function smallestSubsequence(s: string): string {
  let lastOccurence = new Map<string, number>();
  for (let i = 0; i < s.length; i++) {
    lastOccurence.set(s[i], i)
  }

  const seen = new Set<string>();
  const stack = new Array<string>();

  for (let i = 0; i < s.length; i++) {
    const c = s[i];

    if (seen.has(c)) {
      continue;
    }

    while (
      stack.length > 0 &&
      c < stack[stack.length-1] &&
      lastOccurence.get(stack[stack.length-1])! > i
    ) {
        const top = stack.pop()!;
        seen.delete(top);
    }

    stack.push(c);
    seen.add(c);
  }

  return stack.join('');
};
